### DISCUSSION TITLE
beforelink makefile的问题

### DISCUSSION THREAD

**Discussion by DanielKui** [xmake-io/xmake#3907] [Q&A]
Hey，我又来打扰了，有一个问题请教一下，以下是我的.lua脚本，里面有beforelink 拷贝文件操作，但我用
xmake project -k makefile 生成的makefile里，似乎没有cp 这些命令？
![image](https://github.com/xmake-io/xmake/assets/578118/000e01d6-bef1-4c0b-8a6e-34f631dedfcd)
是不是应该有类似这种命令在makefile里

includes ("../xmake.lua")
includes ("../Tools/LogManager/xmake.lua")
includes ("HXQuoteDemoFiles.lua")

add_rules("mode.debug", "mode.release")

target ("HXQuoteApi")
	set_optimize("faster")
	set_languages(cppVersion)
	add_ldflags(linkOptions)
    set_kind ("shared")
	set_targetdir(tarDir)
	add_ldflags("-Wl,-rpath="..tarDir)

    add_defines (ApiDefine)
	add_includedirs(ApiSrcDir)
	add_files(ApiCC)
    add_linkdirs(ApiLibDir)
    add_links(ApiLib)
    
	before_link(function (target)
		os.cp("../Tools/LogManager/bin/libLogManager_x64.so", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libfmt.so*", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libquill.so*", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libprotobuf-lite.so.23", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libACE.so.6.5.9", target:targetdir())
		
		os.cd(target:targetdir())
		os.run("ln -sf libprotobuf-lite.so.23 libprotobuf-lite.so")
		os.cd("..")
		
		os.cd(target:targetdir())
		os.run("ln -sf libACE.so.6.5.9 libACE.so")
		os.cd("..")
    end)
	
	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteApid_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxflags(debugOptions)
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteApi_x64")
		set_optimize("faster")
		add_cxflags(releaseOptions)
	end
	
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

target ("HXQuoteDemo")
	set_kind ("binary")
	set_languages(cppVersion)
	add_ldflags(linkOptions)    
	set_targetdir(tarDir)
    add_defines (DemoDefine)
	
	add_includedirs(DemoSrcDir)
	add_files(DemoCC)
    add_linkdirs(DemoLibDir)
    add_links(DemoLib)
	add_deps("HXQuoteApi")
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteDemod_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxflags(debugOptions)
		add_links("HXQuoteApid_x64")
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteDemo_x64")
		set_optimize("faster")
		add_links("HXQuoteApi_x64")
		add_cxflags(releaseOptions)
	end

盼复

**Comment by waruqi** [comment]
只有 before_linkcmd  on_buildcmd_file 这种 带 cmd 系的脚本才支持 project generator，before_link 是给直接构建用的

**Reply by DanielKui** [reply]
非常感谢， doc  before_linkcmd 的用法能给个链接吗？  刚才在manual 没搜到

**Reply by waruqi** [reply]
源码 rules 目录里 自己搜下

**Reply by waruqi** [reply]
https://github.com/xmake-io/xmake/blob/2c84c686782904a4c13aa4d49d2fe034ba46c4d9/xmake/rules/xcode/metal/xmake.lua#L88

**Comment by DanielKui** [comment]
> 只有 before_linkcmd on_buildcmd_file 这种 带 cmd 系的脚本才支持 project generator，before_link 是给直接构建用的
